### PR TITLE
AP_BLHeli: use correct enum for rover PWM type

### LINKED ISSUE #18729: Rover: brushed bipolar motor is outputting SERVOx_MIN at startup instead of SERVOx_TRIM
This was raised as part of Rover-4.1.0 beta testing here: https://discuss.ardupilot.org/t/brushedbipolar-pwm-problem/76153

### PR DESCRIPTION
**Pull request by IamPete1**
Fixes #18729

The AP motors enum is not the same as the rover one.

This is rover:
https://github.com/ArduPilot/ardupilot/blob/d853d16c42acc21ce762a8db8299926f752b71a4/libraries/AP_Motors/AP_Motors_Class.h#L222-L230

This is AP motors:
https://github.com/ArduPilot/ardupilot/blob/d853d16c42acc21ce762a8db8299926f752b71a4/libraries/AR_Motors/AP_MotorsUGV.h#L15-L25

Rover has `PWM_TYPE_BRUSHED_BIPOLAR` slotted in at 4 so it was showing up as digital when compared the to the AP motors enum. 

The short term fix is to use the right enum. But we really should only have one. This mistake could well exist in other places too, although I could not see any after a very quick look. 

I would guess this also means its impossible to override to `PWM_TYPE_BRUSHED_BIPOLAR` on rover using the `BLH` params.

### CODE REVIEW

**Comment by IamPete1** [comment]
I think I can see another bug, the copters new `PWM_TYPE_PWM_RANGE` will also show up as digital, so will be broken. 

Maybe we should remove `get_pwm_type` in favor of `is_digital_pwm` or something...

File: libraries/AP_BLHeli/AP_BLHeli.cpp [hunk 1]
```diff
@@ -1361,9 +1361,16 @@ void AP_BLHeli::init(void)
      */
 #if APM_BUILD_TYPE(APM_BUILD_Rover)
     AP_MotorsUGV *motors = AP::motors_ugv();
+    if (motors) {
+        uint16_t motormask = motors->get_motor_mask();
```
Comment (andyp1per) [inline libraries/AP_BLHeli/AP_BLHeli.cpp:1365]: This doesn't look really right to me - the otype logic above is wrong for rover still and the masks get fiddled with below again, which also seems wrong. I'm prepared to believe it kind of works but the logic is really hard to follow.

I think you need to add a static function to both AP_Motors and AP_MotorsUGV to do this check and that at least makes this code more logical. We probably also need an AP_Motors::to_rcout_pwm_type() or something to replace the otype logic above.

**Review by andyp1per — CHANGES_REQUESTED** [review]

File: libraries/AP_BLHeli/AP_BLHeli.cpp [hunk 2]
```diff
@@ -1361,9 +1361,16 @@ void AP_BLHeli::init(void)
...
+        // set the rest of the digital channels
+        if (motors->get_pwm_type() >= AP_MotorsUGV::PWM_TYPE_DSHOT150) {
```
Comment (tridge) [inline libraries/AP_BLHeli/AP_BLHeli.cpp:1367]: better to add a using_digital_pwm_type() method?

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by IamPete1** [comment]
Moving the params up will also mean that this is would have shown up in STIL so we would have probably noticed it sooner.

File: libraries/AP_BLHeli/AP_BLHeli.cpp [hunk 3]
```diff
@@ -1315,43 +1315,30 @@ void AP_BLHeli::init(void)
...
-    if (mask && mode != AP_HAL::RCOutput::MODE_PWM_NONE) {
-        hal.rcout->set_output_mode(mask, mode);
-    }
 
     uint16_t digital_mask = 0;
     // setting the digital mask changes the min/max PWM values
     // it's important that this is NOT done for non-digital channels as otherwise
     // PWM min can result in motors turning. set for individual overrides first
-    if (mask && otype >= AP_Motors::PWM_TYPE_DSHOT150) {
+    if (mask && otype >= AP_HAL::RCOutput::MODE_PWM_DSHOT150) {
```
Comment (andyp1per) [inline libraries/AP_BLHeli/AP_BLHeli.cpp:1341]: Need some kind of is_digital() check here

File: ArduCopter/esc_calibration.cpp [hunk 4]
```diff
@@ -153,7 +153,7 @@ void Copter::esc_calibration_setup()
     // clear esc flag for next time
     g.esc_calibrate.set_and_save(ESCCAL_NONE);
 
-    if (motors->get_pwm_type() >= AP_Motors::PWM_TYPE_ONESHOT) {
+    if (motors->is_normal_pwm_type()) {
```
Comment (andyp1per) [inline ArduCopter/esc_calibration.cpp:156]: Your definition of is_normal_pwm_type() is not the same as the code you are replacing

**Review by andyp1per — CHANGES_REQUESTED** [review]

File: ArduCopter/esc_calibration.cpp (hunk 4 above)
Comment (IamPete1) [reply ArduCopter/esc_calibration.cpp:156]: I think the new definition is correct, this case was just overlooked when the range type was added.

File: libraries/AP_BLHeli/AP_BLHeli.cpp (hunk 3 above)
Comment (IamPete1) [reply libraries/AP_BLHeli/AP_BLHeli.cpp:1341]: maybe `RCOutput::is_dshot_protocol(otype)` ?

Comment (IamPete1) [reply libraries/AP_BLHeli/AP_BLHeli.cpp]: I have done that

File: libraries/AP_Motors/AP_Motors_Class.cpp [hunk 5]
```diff
@@ -208,6 +208,21 @@ void AP_Motors::set_limit_flag_pitch_roll_yaw(bool flag)
...
+    switch (_pwm_type) {
+        case PWM_TYPE_DSHOT150:
+        case PWM_TYPE_DSHOT300:
+        case PWM_TYPE_DSHOT600:
+        case PWM_TYPE_DSHOT1200:
+            return true;
+        default:
+            break;
+    }
+    return false;
```
Comment (peterbarker) [inline libraries/AP_Motors/AP_Motors_Class.cpp:223]: We should enumerate all values here so when someone adds a pwm type they're forced to evaluate if it should be added here.

Comment (peterbarker) [reply libraries/AP_Motors/AP_Motors_Class.cpp:223]: Similarly elsewhere.

File: libraries/AP_BLHeli/AP_BLHeli.cpp [hunk 6]
```diff
@@ -1315,43 +1315,30 @@ void AP_BLHeli::init(void)
       allow mode override - this makes it possible to use DShot for
       rovers and subs, plus for quadplane fwd motors
      */
-    AP_HAL::RCOutput::output_mode mode = AP_HAL::RCOutput::MODE_PWM_NONE;
-    AP_Motors::pwm_type otype = AP_Motors::pwm_type(output_type.get());
-
+    // -1 converts from AP_Motors::pwm_type to AP_HAL::RCOutput::output_mode and saves doing a param conversion
+    // this is the only use of the param, but this is still a bit of a hack
+    AP_HAL::RCOutput::output_mode otype = AP_HAL::RCOutput::output_mode(output_type.get() - 1);
```
Comment (tridge) [inline libraries/AP_BLHeli/AP_BLHeli.cpp:1320]: doesn't look like we handle 0 well

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by andyp1per** [comment]
Also needs testing on firmware with and without IOMCU and on plane with mixed dshot outputs

File: libraries/AP_BLHeli/AP_BLHeli.cpp (hunk 6 above)
Comment (IamPete1) [reply libraries/AP_BLHeli/AP_BLHeli.cpp]: added range check

File: libraries/AP_Motors/AP_Motors_Class.cpp (hunk 5 above)
Comment (IamPete1) [reply libraries/AP_Motors/AP_Motors_Class.cpp]: done for the new ones, the moved is_dshot_protocol remains with a default.

**Comment by IamPete1** [comment]
I think this is good to go, I just need to test a few boards IRL as Andy suggests.

**Comment by IamPete1** [comment]
Tested on CubeOrange, all seems to work as expected. Banner reports override type set and digital mask sets min/max/trim to the digital range. I don't have a handy board without a IOMCU to test at the moment. 

Also rebased.
